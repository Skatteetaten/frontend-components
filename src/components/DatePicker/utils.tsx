export const DEFAULTFORMATDATE = (date: Date | null | undefined): string => {
  return date
    ? Intl.DateTimeFormat('no', {
        year: 'numeric',
        month: '2-digit',
        day: 'numeric',
      }).format(date)
    : '';
};

export const DEFAULTPARSEDATEFROMSTRING = (date: string): Date | null => {
  try {
    if (date && date.length === 10) {
      const day: number = parseInt(date.substring(0, 2));
      const month: number = parseInt(date.substring(3, 5));
      const year: number = parseInt(date.substring(6, 10));
      const formattedDate: Date = new Date(Date.UTC(year, month - 1, day));
      return formattedDate;
    }
  } catch (error) {
    console.log(error);
  }
  return null;
};

export const monthsForLocale = (
  localeName = 'no',
  monthFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { month: monthFormat })
    .format;
  const months: string[] = [];
  for (let month = 0; month < 12; month++) {
    months.push(format(new Date(Date.UTC(2020, month))));
  }
  return months;
};

export const weekdaysForLocale = (
  localeName = 'no',
  dayFormat: 'long' | 'short' = 'long'
) => {
  const format = new Intl.DateTimeFormat(localeName, { weekday: dayFormat })
    .format;
  const weekdays: string[] = [];
  for (let day = 0; day < 7; day++) {
    weekdays.push(format(new Date(Date.UTC(2020, 7, day + 2)))); // add 2 days to let the week start on Sunday
  }
  return weekdays;
};
