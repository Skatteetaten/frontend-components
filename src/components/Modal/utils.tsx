import { useEffect } from 'react';

export const createModalDomPlacement = (
  setFlag: (a: boolean) => void,
  shadowRootNode: Document | ShadowRoot | undefined
) => {
  if (!shadowRootNode) {
    setFlag(true);
    return;
  }

  const styleInjectorWrapperId = 'style-injector';
  const modalWrapperId = 'modal-wrapper';

  if (shadowRootNode.getElementById(modalWrapperId)) {
    setFlag(true);
  } else {
    const modalWrapper = document.createElement('div');
    modalWrapper.id = modalWrapperId;
    const styleInjectorWrapper = shadowRootNode.getElementById(
      styleInjectorWrapperId
    );
    const parent = styleInjectorWrapper ? styleInjectorWrapper : shadowRootNode;
    parent.appendChild(modalWrapper);
    setFlag(true);
  }
};

export const useEscOnPress = (
  ref: HTMLDivElement | null,
  onClose?: (ref: HTMLDivElement) => void
): void => {
  useEffect(() => {
    const onKeyDown = (event: KeyboardEvent): void => {
      if (ref == null) {
        return;
      }
      if (event.code === 'Escape' && onClose) {
        onClose(ref);
      }
    };

    if (ref) {
      window.addEventListener('keydown', onKeyDown as EventListener, false);
    }
    return function cleanup(): void {
      if (window) {
        window.removeEventListener('keydown', onKeyDown as EventListener);
      }
    };
  }, [ref, onClose]);
};
