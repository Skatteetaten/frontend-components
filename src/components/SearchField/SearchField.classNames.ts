import { mergeStyles } from '@uifabric/merge-styles';
import { getTheme } from '@uifabric/styling';
import { FontSizes } from '..';
import { PaletteProps } from '..';
// @ts-ignore TODO
export var getClassNames = function getClassNames(props) {
  const palette = getTheme().palette as PaletteProps;
  const { border, searchFieldSize } = props;
  const largeSize = searchFieldSize === 'large';
  const standardSize = searchFieldSize === 'standard';

  return mergeStyles([
    {
      displayName: 'SkeSearchField',
      selectors: {
        '&.ms-SearchBox': {
          borderRadius: '0px',
          height: largeSize ? 48 : 32,
          border:
            border === 'slim'
              ? `1px solid ${palette.skeColor.black}`
              : `2px solid ${palette.skeColor.black}`,
          position: 'relative',
          fontSize: largeSize ? FontSizes.large : FontSizes.medium
        },
        '&.is-active.ms-SearchBox': {
          border: `2px solid ${palette.skeColor.blue}`,
          outline: 'none',
          boxSizing: 'border-box',
          selectors: {
            '.ms-SearchBox-field': {
              padding: border === 'slim' ? '4px 25px 4px 4px' : ''
            },
            '.ms-SearchBox-iconContainer': {
              right: standardSize ? -1 : '',
              top: standardSize ? 6 : ''
            }
          }
        },
        '& .ms-SearchBox-field': {
          outline: 'none',
          fontWeight: 'inherit',
          fontFamily: 'inherit',
          fontSize: 'inherit',
          padding: '5px 30px 5px 5px'
        },
        '& .ms-SearchBox-iconContainer': {
          position: 'absolute',
          right: 0,
          top: largeSize ? 5 : 7
        },
        '& .ms-SearchBox-iconContainer:hover': {
          border: 'none'
        },
        '& .ms-SearchBox-icon': {
          position: 'absolute',
          right: 10,
          top: largeSize ? 5 : 0,
          border: 'none',
          boxSizing: 'border-box',
          fontSize: largeSize ? FontSizes.xLarge : FontSizes.medium,
          opacity: 1,
          outline: 'none'
        },
        '& .ms-SearchBox-clearButton': {
          display: 'none'
        },
        '& .ms-Button': {
          display: 'none'
        }
      }
    }
  ]);
};
