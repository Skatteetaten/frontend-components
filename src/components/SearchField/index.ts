export { SearchField as default } from './SearchField';
export * from './SearchField';
