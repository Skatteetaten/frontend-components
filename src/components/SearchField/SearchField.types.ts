import { IDropdownOption, ISearchBoxProps } from '@fluentui/react';
import { calloutState, LabelWithCalloutProps } from '../LabelWithCallout';
import { Language } from '../utils/Language';

export interface SearchFieldProps extends ISearchBoxProps {
  /** Størrelsen på rammen */
  border?: 'default' | 'slim';
  /** Hjelpetekst */
  help?: string | JSX.Element;
  /** Descriptive label for SearchField */
  label?: string;
  /** aria-label for knapp i label */
  labelButtonAriaLabel?: string;
  /** Overstyr label, se LabelWithCallout komponent */
  labelWithCalloutProps?: LabelWithCalloutProps;
  /** Brukerspesifisert event for callout **/
  onCalloutToggle?: (
    oldCalloutState: calloutState,
    newCalloutState: calloutState
  ) => void;
  /** Liste med mulige søkeresultat fra søk */
  options?: Array<IDropdownOption>;
  /** Størrelsen på søkefeltet */
  searchFieldSize?: 'standard' | 'large';
  /** @ignore */
  underlined?: ISearchBoxProps['underlined'];
  /**påkalt etter bruker valger et alternativ*/
  onSelected?: (option: IDropdownOption) => void;
  /** Language selection for what the screen reader reads out. Default is Norwegian Bokmål */
  language?: Language;
  /** Begrens antall viste søkeresultat */
  limit?: number;
  /** Tillater tastatursnarvei på søk */
  keyboardShortcut?: boolean;
  /** Hvilke taster som fungerer for snarvei */
  searchShortcutKeys?: string;
  /** Gjør søkeikonet klikkbart, trenger samme kode som onSearch */
  onSearchIcon?: (newValue: any) => void;
  /** Legg til egen mouseover tittel på søkeikonet */
  searchIconTitle?: string;
}
