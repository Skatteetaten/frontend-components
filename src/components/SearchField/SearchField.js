import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

import { SearchBox } from 'office-ui-fabric-react/lib-commonjs/SearchBox';

import { getClassNames } from './SearchField.classNames';

/**
 * @visibleName SearchField (Søkefelt)
 */
export default class SearchField extends React.PureComponent {
  static propTypes = {
    /** For ytterligere stiling */
    className: PropTypes.string,
    /** Beskrivende informasjon til skjermlesere */
    ariaLabel: PropTypes.string,
    /** Callback når input endres  */
    onChange: PropTypes.func,
    /** Callback når input fjernes  */
    onClear: PropTypes.func,
    /** Callback når bruker trykker ENTER  */
    onSearch: PropTypes.func,
    /** Tekst inni feltet som vises før man skriver */
    placeholder: PropTypes.string,
    /** Verdi av tekst i søkefeltet */
    value: PropTypes.string,
    /** Størrelsen på søkefeltet */
    searchFieldSize: PropTypes.oneOf(['standard', 'large']),
    /** Størrelsen på rammen */
    border: PropTypes.oneOf(['default', 'slim']),
    /** Global attributt som må være unik for hele HTML dokumentet */
    id: PropTypes.string
  };

  static defaultProps = {
    searchFieldSize: 'standard',
    border: 'default'
  };

  render() {
    const { searchFieldSize, border, className, id, ...props } = this.props;

    return (
      <div id={id}>
        <SearchBox
          {...props}
          className={classnames(getClassNames(this.props), className)}
          searchFieldSize={searchFieldSize}
        />
      </div>
    );
  }
}
