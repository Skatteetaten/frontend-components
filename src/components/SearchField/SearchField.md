** SearchField er et tekstfelt som brukes når man søker gjennom innhold i en nettløsning. **

```js
import SearchField from '@skatteetaten/frontend-components/SearchField';

<div style={{ width: '300px' }}>
  <SearchField
    searchFieldSize="standard"
    border="slim"
    placeholder="Skriv søkeord her"
    ariaLabel="Søk"
    keyboardShortcut={true}
    onSearchIcon={() => alert('Du trykket på søkeikonet')}
    searchIconTitle="Start søk"
  />
  <br />
  <SearchField
    searchFieldSize="large"
    placeholder="Skriv søkeord her"
    ariaLabel="Søk"
  />
</div>;
```

Søkefelt med resultat i dropdown:

```js
import SearchField from '@skatteetaten/frontend-components/SearchField';
const options = [
  { key: '1', text: 'Banan' },
  { key: '2', text: 'Ananas' },
  { key: '3', text: 'Eple' },
  { key: '4', text: 'Pære' },
  { key: '5', text: 'Appelsin' },
  { key: '6', text: 'Fersken' }
];

<div style={{ width: '300px' }}>
  <SearchField
    label={'Søk etter tilgjengelige frukter'}
    searchFieldSize="standard"
    border="slim"
    placeholder="Skriv søkeord her"
    ariaLabel="Søk"
    help="Tekst som hjelper brukeren å fylle ut feltet."
    options={options}
  />
</div>;
```

```js noeditor beskrivelse
<>
  <h3>Markeres med ikon eller tekst</h3>
  <p>
    Et søkefelt skal alltid ha en markering i form av et søkeikon eller
    alternativt en knapp med «Søk».
  </p>
  <h3>Plassering og bredde</h3>
  <p>
    Søkefeltet skal kun ta en del av bredden på skjermen eller området for
    tekst. Hvis feltet blir for stort, oppfatter brukeren det som en banner.
  </p>
  <p>
    Søkefeltet kan legges inni en egen container, og størrelsen kan endres
    deretter.
  </p>
  <h3>Tastatursnarvei</h3>
  <p>
    SearchField tillater tastatursnarvei for søk (ctrl+f eller command+f). For å
    slå dette på legg til prop keyboardShortcut=&#123;true&#125;
  </p>
</>
```

```js noeditor uu
<>
  <h3>Tips</h3>
  <ul>
    <li>
      Vanlige søkefelt skal ha tabstopp på søkeknapp (forstørrelsesglasset).
      Unntaket er søkefelt med forslag i dropdown.
    </li>
    <li>
      For søkefelt med synlig ledetekst, brukes input med label for-attributt.
      Søkefelt uten synlig ledetekst kan bruke aria-label på input.
    </li>
    <li>
      Kun placeholder-attributt er ikke tilstrekkelig for å bestå WCAG.
      Ledetekst (label/aria-label) og placeholder bør ikke være identiske.
      Placeholder-tekst er ment for å gi et hint om det som skal skrives i
      feltet.
    </li>
  </ul>

  <h3>Mest relevante WCAG-krav</h3>
  <ul>
    <li>1.3.1 A, Informasjon og relasjoner </li>
    <li>3.3.2 A, Ledetekster eller instruksjoner</li>
    <li>4.1.2 A, Navn, Rolle, verdi</li>
  </ul>

  <h3>WAI-ARIA</h3>
  <ul>
    <li>Aria-label brukes på søkeknapp for å navngi den for skjermlesere.</li>
  </ul>
</>
```
