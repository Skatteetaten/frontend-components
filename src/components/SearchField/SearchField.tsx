import classnames from 'classnames';
import {
  ISearchBoxProps,
  SearchBox
} from 'office-ui-fabric-react/lib-commonjs/SearchBox';
import * as React from 'react';
import { getClassNames } from './SearchField.classNames';

interface SearchFieldProps extends ISearchBoxProps {
  /** Størrelsen på søkefeltet */
  searchFieldSize?: 'standard' | 'large';
  /** Størrelsen på rammen */
  border?: 'default' | 'slim';
}

/**
 * @visibleName SearchField (Søkefelt)
 */
export default class SearchField extends React.PureComponent<
  SearchFieldProps,
  {}
> {
  static defaultProps = {
    border: 'default',
    searchFieldSize: 'standard'
  };

  render() {
    const { className, ...props } = this.props;

    return (
      <SearchBox
        {...props}
        className={classnames(getClassNames(this.props), className)}
      />
    );
  }
}
