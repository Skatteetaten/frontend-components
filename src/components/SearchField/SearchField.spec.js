import React from 'react';
import { shallow, mount } from 'enzyme';
import { matches } from './../utils/test-utils';
import SearchField from './SearchField';

function oppsettShallow(props) {
  return shallow(<SearchField {...props} />);
}

function oppsettFullDOM(props) {
  return mount(<SearchField {...props} />);
}

describe('SearchField komponent', () => {
  it('matcher snapshot', () => {
    matches(<SearchField />);
  });

  it('rendrer SearchField med default props ', () => {
    const wrapper = oppsettShallow({});
    expect(wrapper.find('StyledSearchBoxBase').prop('searchFieldSize')).toEqual(
      'standard'
    );
  });

  it('rendrer SearchField med riktige props ', () => {
    const wrapper = oppsettFullDOM({
      value: 'searchfield-value',
      ariaLabel: 'searchfield-label',
      placeholder: 'searchfield-placeholder',
      className: 'searchfield-classname',
      id: 'searchfield-id',
      searchFieldSize: 'large'
    });

    const searchField = wrapper.find('input.ms-SearchBox-field');
    const StyledSearchBoxBase = wrapper.find('StyledSearchBoxBase');
    expect(wrapper.first().prop('id')).toEqual('searchfield-id');
    expect(StyledSearchBoxBase.prop('className')).toContain(
      'searchfield-classname'
    );
    expect(StyledSearchBoxBase.prop('ariaLabel')).toEqual('searchfield-label');
    expect(StyledSearchBoxBase.prop('searchFieldSize')).toEqual('large');
    expect(searchField.prop('placeholder')).toEqual('searchfield-placeholder');
    expect(searchField.prop('value')).toEqual('searchfield-value');
  });
});
