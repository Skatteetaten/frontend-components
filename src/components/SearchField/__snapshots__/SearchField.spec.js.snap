// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField komponent matcher snapshot 1`] = `
<div
  className="ms-SearchBox SkeSearchField-41"
  onFocusCapture={[Function]}
  role="search"
>
  <div
    aria-hidden={true}
    className="ms-SearchBox-iconContainer iconContainer-42"
    onClick={[Function]}
  >
    <i
      aria-hidden={true}
      className="ms-SearchBox-icon icon-47"
      data-icon-name="Search"
    >
      Ôçâ
    </i>
  </div>
  <input
    className="ms-SearchBox-field field-45"
    id="SearchBox0"
    onChange={[Function]}
    onInput={[Function]}
    onKeyDown={[Function]}
    role="searchbox"
    value=""
  />
</div>
`;
